### PR TITLE
[1953] Harden LLM Chat correctness, security, and accessibility

### LINKED ISSUE #1953: Harden LLM Chat correctness, security, and accessibility
## Description

Harden the LLM Chat implementation before it becomes the Governance Playground. Improve type safety, input validation, cancellation, rendering, and accessibility.

## Expected Behavior

- Inference and MCP responses use explicit TypeScript types.
- Empty prompts and invalid model or tool setup cannot start a request.
- Invalid extra-parameter JSON produces an inline validation message.
- `AbortSignal` reaches inference and tool-execution requests.
- Cancellation stops execution predictably.
- Multi-part and non-text MCP results render correctly.
- Tool responses do not use unsafe HTML rendering.
- Tool and namespace selection remains internally consistent.
- Transcript markup and execution status are accessible to screen readers.
- Playwright tests cover errors, cancellation, keyboard operation, and rendering.

## Additional Context

The selected-namespace tool-list bug is already tracked by #1932. Fix or depend on that issue instead of duplicating it in this scope.

### PR DESCRIPTION
**Pull request by rnetuka**
Issue: https://github.com/wanaku-ai/wanaku/issues/1953

- Inference and MCP responses use explicit TypeScript types.
  > @opiske not sure what you meant
- Empty prompts and invalid model or tool setup cannot start a request.
  > fixed
- Invalid extra-parameter JSON produces an inline validation message.
  > fixed
- AbortSignal reaches inference and tool-execution requests.
  > fixed
- Cancellation stops execution predictably.
  > @opiske not sure what you meant
- Multi-part and non-text MCP results render correctly.
  > @opiske not sure about multi-part, but I 've added non-text responses. Currently displays only the type, later we could    improve it by displaying the image, audio etc.
- Tool responses do not use unsafe HTML rendering.
  > @opiske should alread work. If not, please provide an example
- Tool and namespace selection remains internally consistent.
  > @opiske should lready work. There was a problem that namespace stored if the local storage might not be valid anymore. Selecting a different one should help fix it. Alternatively, try to delete the local storage.
- Transcript markup and execution status are accessible to screen readers.
  > @opiske not sure what you meant
- Playwright tests cover errors, cancellation, keyboard operation, and rendering.
  > @opiske not sure what you meant

## Summary by Sourcery

Harden the LLM chat experience by validating request configuration, handling cancellation and MCP responses safely, and keeping tool selection state consistent.

New Features:
- Validate selected models against the inference service before sending chat requests.
- Support non-text MCP tool results with type-aware transcript placeholders.

Bug Fixes:
- Prevent invalid chat requests caused by empty prompts, missing credentials or models, invalid model selections, and malformed extra parameters.
- Keep tool namespace and select-all behavior synchronized with the currently available tools.

Enhancements:
- Propagate request cancellation signals to inference calls and improve send-button state management.
- Surface chat validation and execution errors through the page-level error display.

### CODE REVIEW

File: ui/admin/src/Pages/LLMChat/extra-llm-parameters.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  const parameters = JSON.parse(extraLlmParameters) as Record<string, unknown>
+  
+  if ("top_k" in parameters) {
+    assertIsInteger("top_k", parameters)
+    const top_k = parameters.top_k as number
+    if (top_k < 0 && top_k !== -1) {
+      throw new RangeError("Parameter 'top_k' must be positive number, greater than zero");
+    }
+  }
+  assertNumberIfPresent("top_p", parameters)
```
Comment (orpiske) [inline ui/admin/src/Pages/LLMChat/extra-llm-parameters.ts:22]: The condition `top_k < 0 && top_k !== -1` is unreachable for `-1`. Since `-1 < 0` is true, the `&&` short-circuits to check `top_k !== -1`, which is false when `top_k === -1`. This means the sentinel value `-1` (commonly used for "unlimited") throws an error instead of being accepted.

Suggestion:
```suggestion
    if ((top_k < 0 || top_k > 1) && top_k !== -1) {
      throw new RangeError("Parameter 'top_k' must be positive number, greater than zero");
    }
```

**Review by orpiske — COMMENTED** [review]
# PR #1975 Review — Harden LLM Chat correctness, security, and accessibility

## Confirmed Issues

### 1. Missing tests for new modules (Severity: Medium)
`extra-llm-parameters.ts` (127 lines of validation logic) and `use-models.ts` have zero test coverage. The `validateExtraLlmParameters` function handles many parameter types and edge cases; any regression here silently affects all LLM chat requests. Unit tests covering each assertion helper and the full validation path are recommended.

### 2. Stale model cache (Severity: Low)
The `listModels` function caches results in a module-level variable forever. If inference endpoints change (new models added/removed), the UI will serve stale data until a hard page refresh. Consider adding a simple TTL (e.g., 5 minutes) or exposing an invalidate function.

See: [use-models.ts:14](ui/admin/src/hooks/api/use-models.ts#L14)

### 3. `top_k` validation boundary bug (Severity: Medium)
The condition `if (top_k < 0 && top_k !== -1)` has contradictory logic. Since `-1 < 0` is true but `-1 !== -1` is false, the entire expression evaluates to `false` when `top_k === -1`. This means the sentinel value `-1` (commonly used for "unlimited" in LLM APIs) will throw an error instead of being accepted. The correct logic should allow `[0, 1]` and `-1`.

See: [extra-llm-parameters.ts:22](ui/admin/src/Pages/LLMChat/extra-llm-parameters.ts#L22)

### 4. Missing newline at EOF (Severity: Trivial)
`extra-llm-parameters.ts` lacks a trailing newline. Minor style violation.

## Questions / Assumptions

### 5. Model validation on every send
`validateSelectedModel()` calls the inference API on each send. For repeated sends this adds latency. Is this intentional, or should validation happen once at mount/config-change time?

### 6. No Playwright E2E tests
Issue #1953 explicitly calls for "Playwright tests cover errors, cancellation, keyboard operation, and rendering." None were added. Is this out of scope for this PR, or planned separately?

## Positive Notes

- `LLMTools.tsx` select-all fix correctly addresses the filtered namespace selection bug.
- AbortSignal propagation to inference fetch improves predictable cancellation.
- Non-text MCP content type mapping prevents unsafe HTML rendering.
- Send button state management provides better UX feedback.

---

This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying.

File: ui/admin/src/hooks/api/use-models.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+interface ModelCache {
+  models: string[]
+}
+
+let cache: ModelCache | null = null
+
+
+export async function listModels(apiKey: string): Promise<string[]> {
+  if (cache) {
+    return cache.models
```
Comment (orpiske) [inline ui/admin/src/hooks/api/use-models.ts:13]: Module-level cache persists forever. If inference endpoints change (new models added/removed), the UI serves stale data until a hard refresh. Consider adding a TTL or invalidate mechanism.

File: ui/admin/src/Pages/LLMChat/LLMChatArea.tsx [hunk 3]
```diff
@@ -116,7 +147,19 @@ export const LLMChatArea: React.FC<LLMChatAreaProps> = ({ config, onSystemPrompt
                   name: toolName,
                   arguments: toolArgs
                 })
-                const toolResultText = (toolResult.content as Array<{ text: string }>)[0].text
+                const toolResultType = (toolResult.content as any).type
```
Comment (orpiske) [inline ui/admin/src/Pages/LLMChat/LLMChatArea.tsx:150]: Nice addition — non-text MCP content types now render as safe placeholders instead of potentially unsafe HTML.

File: ui/admin/src/Pages/LLMChat/LLMChatArea.tsx [hunk 4]
```diff
@@ -50,15 +54,42 @@ export const LLMChatArea: React.FC<LLMChatAreaProps> = ({ config, onSystemPrompt
...
+       throw new Error("Selected model is not valid")
+    }
+    validateExtraLlmParameters(config.extraLlmParams)
     try {
       chatHistory.current.push({ role: "user", content: userPrompt })
       setDisplayedMessages(chatHistory.current)
       setIsRunning(true)
+      setSendDisabled(true)
       
       async function send(): Promise<Response> {
```
Comment (orpiske) [inline ui/admin/src/Pages/LLMChat/LLMChatArea.tsx:89]: AbortSignal propagation to fetch is a solid improvement for predictable cancellation.

File: ui/admin/src/Pages/LLMChat/LLMTools.tsx [hunk 5]
```diff
@@ -119,7 +119,7 @@ export const LLMTools: React.FC<LLMToolsProps> = ({
             checked={isAllSelected()}
             indeterminate={isSomeSelected()}
             onChange={(_, { checked }) => {
-              const selection = checked ? [...tools] : []
+              const selection = checked ? [...filteredTools()] : []
```
Comment (orpiske) [inline ui/admin/src/Pages/LLMChat/LLMTools.tsx:122]: Good fix — select-all now respects the active filter instead of selecting all tools regardless of visibility.

**Comment by orpiske** [comment]
Looking better after the second round of fixes. Please, can you also adjust so that it updates the list of tools after I select a namespace? Right now I have to select the namespace and then refresh the page so it updates the list of tools for the selected namespace.

**Comment by orpiske** [comment]
@rnetuka looking better now. Still some rendering issues when calling tools, though.

Screenshot 2026-09-15 at 10 39 54

**Comment by orpiske** [comment]
@rnetuka anything else on this one? Should we merge and leave the tool call rendering issues for later? Just let me know .... the core issues are fixed, IMHO, we should be ready to merge.

**Review by orpiske — APPROVED** [review]
Ok, let's get this one merged
